### PR TITLE
Improve Regexp escaping

### LINKED ISSUE #1455: Invalid Regexp escaping in Opal 0.10.0.beta3 with Node.js
Given the following Ruby code:

```ruby
CommentBlockRx = %r{^/{4,}$}
```

https://github.com/asciidoctor/asciidoctor/blob/v1.5.4/lib/asciidoctor.rb#L586

Here's the result:
**Opal 0.9.2**:
```javascript
Opal.cdecl($scope, 'CommentBlockRx', /^\/{4,}$/);
```

**Opal 0.10.0.beta3**:
```javascript
Opal.cdecl($scope, 'CommentBlockRx', /^/{4,}$/);
```

You can see that the second front slash is not escaped.

### PR DESCRIPTION
**Pull request by Mogztter**
Resolves #1455

### CODE REVIEW

**Comment by wied03** [comment]
@Mogztter - There should probably be rubyspecs to deal with this, right?

Maybe one of these?
```ruby
fails "Regexp#inspect correctly escapes forward slashes /"
  fails "Regexp#inspect does not include a character set code"
  fails "Regexp#inspect does not include the 'o' option"
  fails "Regexp#inspect does not over escape"
  fails "Regexp#inspect escapes 2 slashes in a row properly"
```

**Comment by Mogztter** [comment]
This is not working for regexp with multiples front slashes. A good example is: `^//(?:[^\/]|$)`.
I tried various regexp combinaison but none are fully working :cry:

**Comment by Mogztter** [comment]
Flags is not supported in V8... https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/String/replace

I've decided to replace the regexp with a loop over the characters.

@wied03 The following rspec are now enabled :tada: 

 * [x] "Regexp#inspect correctly escapes forward slashes /"
 * [x]  "Regexp#inspect does not over escape"
 * [x]  "Regexp#inspect escapes 2 slashes in a row properly"

**Comment by wied03** [comment]
👍

File: opal/corelib/regexp.rb [hunk 1]
```diff
@@ -118,7 +118,36 @@ def =~(string)
   alias eql? ==
 
   def inspect
-    `self.toString()`
+    # Use a regexp to extract the regular expression and the optional mode modifiers from the string.
+    # In the regular expression, escape any front slash (not already escaped) with a backslash.
+    %x{
+      var regexp_format = /^\/(.*)\/(.*)/;
```
Comment (elia) [inline opal/corelib/regexp.rb]: alt proposal: `/^\/(.*)\/([^\/]*)$/`

**Comment by elia** [comment]
@Mogztter marked as WIP, ping when the task list is completed (I see `"Regexp#inspect does not over escape"` is still to be done)

**Comment by Mogztter** [comment]
Yes this one is tricky. I will try your regexp ;)

**Comment by Mogztter** [comment]
> marked as WIP, ping when the task list is completed 

@elia I'm stuck with "Regepx#inspect does not over escape" spec. I've tried at least a dozen Regexp variation but none worked. I've also tried your regexp:
```
Regexp#inspect does not over escape FAILED
Expected "/\\\\//" to equal "/\\\\\\//"
SpecExpectationNotMetError: Expected "/\\\\//" to equal "/\\\\\\//"
```

I think this is still an improvement compare to the current state because we are now passing two more specs. Can we leave "Regepx#inspect does not over escape" spec aside for now ?

File: .gitignore [hunk 2]
```diff
@@ -14,6 +14,7 @@ build/
 .rvmrc
 .rspec-local
 /coverage
+.idea/
```
Comment (elia) [inline .gitignore]: this should go in your global .gitignore, see instructions here https://help.github.com/articles/ignoring-files/#create-a-global-gitignore 🙂

File: opal/corelib/regexp.rb [hunk 3]
```diff
@@ -118,7 +118,36 @@ def =~(string)
...
+    # In the regular expression, escape any front slash (not already escaped) with a backslash.
+    %x{
+      var regexp_format = /^\/(.*)\/([^\/]*)$/;
+      var value = self.toString();
+      var matches = regexp_format.exec(value);
+      if (matches) {
+        var regexp_pattern = matches[1];
+        var regexp_flags = matches[2];
+        var chars = regexp_pattern.split('');
+        var charsLength = chars.length;
```
Comment (elia) [inline opal/corelib/regexp.rb]: please use snake_case here: `chars_length`

**Comment by elia** [comment]
@Mogztter started digging but then realized it seems to be fixed on master:

```
opal -v -ropal-parser -e 'puts Opal.compile("CommentBlockRx = %r{^/{4,}$}")'                                                                                ~/C/opal
Opal v0.10.0.beta5
/* Generated by Opal 0.10.0.beta5 */
(function(Opal) {
  var self = Opal.top, $scope = Opal, nil = Opal.nil, $breaker = Opal.breaker, $slice = Opal.slice;

  return Opal.cdecl($scope, 'CommentBlockRx', /^\/{4,}$/)
})(Opal);
```

Please give 0.10.0.beta5 a try and let me know if I'm missing something.
If all is good I'll merge just the added spec from this PR to avoid regressions. :)

**Comment by Mogztter** [comment]
Wow beta5 already, I missed beta4! OK I will give it a try.

Thanks for your link on global gitignore, I didn't know about it.

**Comment by elia** [comment]
> Wow beta5 already, I missed beta4! OK I will give it a try.

`beta4` was promptly replaced by `beta5` because of a rubygems bug that made it difficult to install with some rubygems versions. `beta5` was an attempt to overcome the bug, but I already got feedback that's still present.

**Comment by Mogztter** [comment]
@elia Just tried with master and I got the following exception:

```javascript
Opal.cdecl($scope, 'CommentBlockRx', /^/{4,}$/);
                                            ^
SyntaxError: Unexpected token {
    at exports.runInThisContext (vm.js:73:16)
    at Module._compile (module.js:443:25)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
```
Here's the Ruby code: https://github.com/asciidoctor/asciidoctor/blob/v1.5.4/lib/asciidoctor.rb#L586
```ruby
CommentBlockRx = %r{^/{4,}$}
```

**Comment by Mogztter** [comment]
@elia Hmmm I don't know if my local environment is dirty but here is my output with `master` and `regexp-escape` branch:

```
~/workspace/opensource/opal@master: 
$ opal -v -ropal-parser -e 'puts Opal.compile("CommentBlockRx = %r{^/{4,}$}")'
Opal v0.10.0.beta5
```
```javascript
/* Generated by Opal 0.10.0.beta5 */
(function(Opal) {
  var self = Opal.top, $scope = Opal, nil = Opal.nil, $breaker = Opal.breaker, $slice = Opal.slice;

  return Opal.cdecl($scope, 'CommentBlockRx', /^/{4,}$/)
})(Opal);
```
```
~/workspace/opensource/opal@master: 
$ git checkout -
Switched to branch 'regexp-escape'
```
```
~/workspace/opensource/opal@regexp-escape
$ opal -v -ropal-parser -e 'puts Opal.compile("CommentBlockRx = %r{^/{4,}$}")'
Opal v0.10.0.beta5
```
```javascript
/* Generated by Opal 0.10.0.beta5 */
(function(Opal) {
  var self = Opal.top, $scope = Opal, nil = Opal.nil, $breaker = Opal.breaker, $slice = Opal.slice;

  return Opal.cdecl($scope, 'CommentBlockRx', /^\/{4,}$/)
})(Opal);
```

**Comment by Mogztter** [comment]
I will try with 0.10.0.rc1

**Comment by Mogztter** [comment]
Happy to announce that `Regexp#inspect does not over escape` is finally working :+1: 
@elia Let me know if this looks good to you.

**Comment by Mogztter** [comment]
I think we can remove "work-in-progress" label.
By the way this is the last issue before I can compile Asciidoctor to JavaScript using `opal-builder.js` :smile: 

@iliabylich Could you please review this pull request ? Thanks

**Comment by Mogztter** [comment]
Thanks! Let's celebrate! :tada: :beers:

**Comment by elia** [comment]
🍾🍷😉

Elia Schito

**Pull request merged**